### PR TITLE
Implemented Save  Page

### LINKED ISSUE #77: Implement "Save page"
It is not possible to save pages in Kiwix now. When you click "Save page" nothing happens.

### PR DESCRIPTION
**Pull request by ShaopengLin**
Fix #77 
Changes:
- Re-enabled file menu item and shortcut Ctrl + Shift + S for save page.
- Saves the HTML of the page on the current tab to PDF.
- Saves non-HTML pages directly as files.
- Remembers the last directory saved when opening file selection dialoge.

### CODE REVIEW

File: src/webview.cpp
Comment (ShaopengLin) [inline src/webview.cpp:158]: Do we really need a `gt` translation here? The filter seems like it should be a pretty standard translation.

File: src/webview.cpp
Comment (ShaopengLin) [inline src/webview.cpp:158]: Is the `gt` for the filter text needed? I didn't see one for our .zim filters but I added it just in case.

**Comment by ShaopengLin** [comment]
Actually, we should hold off until #1132 is merged as we have to test how that works. I am also seeing some places of improvement for this PR.

**Comment by kelson42** [comment]
@ShaopengLin I will review this PR, can you please rebase and fix the conflict?

**Comment by ShaopengLin** [comment]
@kelson42 done

**Comment by kelson42** [comment]
* Why `CTRL + shift + S`? usual (Web Browser) shortcut is `CTRL + S`, at least on linux. (otherwise it works)
*

**Comment by ShaopengLin** [comment]
> If I want to save the article "AC/DC" then it tries to create DC.pdf

@kelson42 I may be wrong I believe your case happens like this: The native/Qt window manager would treat it as if AC is a directory and DC.pdf is the name. The path will be path/AC/DC.pdf which simply fails to save. For this case, there is no way for us to know if the user intended for it to be a path element or a file name when the results get back to us. I don't think we have a choice here but just report an error.

In most cases, native Linux window manager simply doesn't allow this to happen. I don't believe windows' window manager will allow saves to invalid file names either.

**Comment by ShaopengLin** [comment]
@kelson42 Done.

> Saving a mp4 video directly does not work, it proposes to save a PDF file, not directly the .mp4 file.

The video/contents that can be saved are these ones opened directly as a video file in the tab:
Screenshot from 2024-07-15 01-14-42

The ones embedded inside a html web page would we would not be able to identify those as the page mime type is html:
Screenshot from 2024-07-15 01-15-02

These videos can just be downloaded using the provided download button:
Screenshot from 2024-07-15 01-15-11

**Comment by kelson42** [comment]
@ShaopengLin  Please update me on each point I have given. In general this is verz imoortant to be verz clear about what is done and why. The reviewer have a limited amount of time and therefore the goal is to implement PR so the reviewer as just to approve it without having to ask questions and obviouslz ask for changes.

**Comment by ShaopengLin** [comment]
@kelson42
1. The shortcut for save is now Ctrl + S. Qt's "Save As" shortcut Ctrl + Shift + S assumes a desktop environment instead of a browser one.
2. The issue is explained in my previous [comment](https://github.com/kiwix/kiwix-desktop/pull/1134#issuecomment-2227718729).  
3. We now do a simple sanitization on the appearances of reserved characters of the OS in the file name by replacing them with '_'. This prevents our code from parsing a file name "AC/DC" as file "DC" in the directory "AC". All we care about here is preventing any side effects from using the title text as file name, as it should just be a placeholder name. Determining the file name validity is the job of the OS file window manager.

File: src/webview.cpp
Comment (kelson42) [reply src/webview.cpp]: To me it seems mandatory, otherwise how would you get a translated message?! Or I get it wrong?! @veloman-yunkan

**Comment by kelson42** [comment]
I have rebased as we are approaching the time to make the code review and the PR is already quite old.

**Comment by kelson42** [comment]
@veloman-yunkan @ShaopengLin I would recommend to move the string slugification method to libkiwix `stringTools.cpp`. What do you think?

**Comment by kelson42** [comment]
@ShaopengLin Looks good to me finally. Could you please add as well the the contextual menu like on a Web Browser?
@veloman-yunkan Ready to code review.

**Review by kelson42 — COMMENTED** [review]
LGTM

**Comment by ShaopengLin** [comment]
I would be fine with moving it to libkiwix. Note this doesn't guarantee the file name is 100% usable in Windows. They have a lot of rules extra rules.

**Comment by ShaopengLin** [comment]
@kelson42 Do you want me to just PR from fork to libkiwix, or create a branch like we do here? If the latter I will need to be a collaborator.

**Comment by kelson42** [comment]
> @kelson42 Do you want me to just PR from fork to libkiwix, or create a branch like we do here? If the latter I will need to be a collaborator.

Sorry, you should just have received an invitation to be member of the team and have write permission.

**Comment by kelson42** [comment]
Blocked by https://github.com/kiwix/libkiwix/pull/1105

**Comment by ShaopengLin** [comment]
@kelson42 @veloman-yunkan Ready for review. Updated change from https://github.com/kiwix/libkiwix/pull/1105.

**Comment by kelson42** [comment]
@ShaopengLin I want to test it, but would be very nice if you could rebase it (once again).

**Comment by ShaopengLin** [comment]
@kelson42 Done.

**Review by veloman-yunkan — CHANGES_REQUESTED** [review]

File: src/kiwixapp.cpp
Comment (ShaopengLin) [reply src/kiwixapp.cpp:219]: Forgot to include it when doing fixes. @veloman-yunkan if we do not care about the overhead of reading this directory from session for each `getPrevSaveDir()`, then we can merge them.

File: src/library.cpp
Comment (ShaopengLin) [reply src/library.cpp:85]: @veloman-yunkan I think making this a static member makes more sense. Having getZimIdFromUrl is helpful, but it shouldn't belong to `Library`.

File: src/kiwixapp.cpp
Comment (veloman-yunkan) [reply src/kiwixapp.cpp:219]: I see no reason why we should mind about that overhead

File: src/library.cpp
Comment (ShaopengLin) [reply src/library.cpp:85]: Yeah, it does seem this doesn't belong to Library either. Might be better suited as a free function in library.h/cpp

File: resources/i18n/en.json
Comment (ShaopengLin) [reply resources/i18n/en.json:185]: Don't think the text should change as Chromium browsers also just say "Save As..."

Comment (ShaopengLin) [reply resources/i18n/en.json]: Done

File: src/kiwixapp.cpp
Comment (ShaopengLin) [reply src/kiwixapp.cpp]: Done

File: src/library.cpp
Comment (veloman-yunkan) [inline src/library.cpp:26]: I think you can preserve this function in `src/urlschemehandler.cpp` as an enhancement of `getEntryFromPath()` (this way the change will be more conspicuous).

File: src/library.h
Comment (veloman-yunkan) [inline src/library.h:65]: Declaring functions in header files is convenient but not mandatory. You can simply forward declare it in the second .cpp file (accompanied by a comment pointing to the location of the definition) and do that in the commit when that second usage is introduced. This way the issue of finding an appropriate header file for this declaration (and the ensuing sense of necessity of moving the definition in the respective cpp file) is worked around.

File: resources/i18n/qqq.json
Comment (veloman-yunkan) [inline resources/i18n/qqq.json:192]: pagr → you know what

File: src/kprofile.cpp
Comment (veloman-yunkan) [inline src/kprofile.cpp]: This better be done in the previous (or separate) commit

**Review by veloman-yunkan — CHANGES_REQUESTED** [review]
If it were not for the pagr typo I might have approved. But since we are going to go through another iteration I chose not to silence the perfectionist side of me :)

File: src/kprofile.cpp
Comment (ShaopengLin) [reply src/kprofile.cpp]: Done

File: resources/i18n/qqq.json
Comment (ShaopengLin) [reply resources/i18n/qqq.json]: Done

File: src/library.h
Comment (ShaopengLin) [reply src/library.h]: Done

File: src/library.cpp
Comment (ShaopengLin) [reply src/library.cpp]: Done

File: src/urlschemehandler.cpp
Comment (veloman-yunkan) [inline src/urlschemehandler.cpp:34]: The placement of `getArchiveEntryFromUrl()` and the change in whitespace make the diff in this commit more confusing than it could be. History of code is as important as its most recent state. Please don't introduce unnecessary difficulties in understanding the changes in the code. Cleanup actions like moving code around for better cohesion etc are welcome but they should be performed as separate steps in commits of their own. A general rule - try to not combine changes of different nature in the same commit and try keep the size of each change within an average software engineer's ability to comprehend it in a single pass.

**Review by veloman-yunkan — CHANGES_REQUESTED** [review]
The commit that dropped category translations was lost. Was that a mistake?

File: src/urlschemehandler.cpp
Comment (ShaopengLin) [reply src/urlschemehandler.cpp]: Done, the change log does look confusing...

**Review by veloman-yunkan — APPROVED** [review]

**Pull request merged**
